### PR TITLE
PolyExponential, Spline and DMD ROMs

### LINKED ISSUE #627: Addition of intrinsic time-dependent regressors
--------
Issue Description
--------

**New feature** 

We would need to add new surrogates that are intrinsically able to handle time-dependency.
2 surrogate models are suggested:
- a sum of exponential-based SM
- dynamic mode decomposition

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by alfoa**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #627

This PR adds two new surrogate model (for time-dependent) that are aimed to treat the
time-dependency explicitly:
* Dynamic Mode Decomposition
* PolyExponential

In addition, it fixes the NDSpline ROM that now can be used ad regressor. 

##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

File: framework/SupervisedLearning.py [hunk 1]
```diff
@@ -2823,6 +2869,518 @@ def reseed(self,seed):
...
+    if len(self.target) > 2:
+      self.raiseAnError(IOError,"Multi-target PolyExponential not available yet!")
+    self.targetID = self.target[self.target.index(self.pivotParameterID) - 1]
+    if self.polyExpParams['coeffRegressor'] not in self.availCoeffRegressors:
+      self.raiseAnError(IOError, ' Only the following "coeffRegressor" are available: ' +",".join(self.availCoeffRegressors))
+
+  def _localNormalizeData(self,values,names,feat):
+    """
+      Overwrites default normalization procedure.
+      @ In, values, unused
```
Comment (mandd) [inline framework/SupervisedLearning.py]: do we need to have them defined if they are not used?

File: framework/SupervisedLearning.py [hunk 2]
```diff
@@ -2823,6 +2869,518 @@ def reseed(self,seed):
...
+      @ In, feat, feature to normalize
+      @ Out, None
+    """
+    self.muAndSigmaFeatures[feat] = (0.0,1.0)
+
+  def __computeExponentialTerms(self, x, y, returnPredictDiff=False):
+    """
+      Method to compute the coefficients of "n" exponential terms that minimize the
+      difference between the training data and the "predicted" data
+      y(x) = \sum_{i=1}^n a_i \exp ( - b_i x )
```
Comment (mandd) [inline framework/SupervisedLearning.py]: I see underscores here!!!! :-P just kidding

File: framework/SupervisedLearning.py [hunk 3]
```diff
@@ -2823,6 +2869,518 @@ def reseed(self,seed):
...
+        @ In, s, numpy.ndarray, the array of coefficient
+        @ Out, objective, float, the cumulative difference between the predicted and the real data
+      """
+      l = int(s.size/2)
+      return np.sum((y - np.dot(s[l:], np.exp(-np.outer(1./s[:l], x))))**2.)
+
+    predictionErr = None
+    x, y          = np.array(x), np.array(y)
+    bounds        = [[min(x), max(x)]]*self.polyExpParams['expTerms'] + [[min(y), max(y)]]*self.polyExpParams['expTerms']
+    result        = differential_evolution(_objective, bounds, maxiter=self.polyExpParams['maxNumberIter'], tol=self.polyExpParams['tol'],disp=False)
```
Comment (mandd) [inline framework/SupervisedLearning.py]: differential_evolution is a scipy method, right?

Comment (mandd) [reply framework/SupervisedLearning.py]: yes, just checking

File: framework/SupervisedLearning.py [hunk 4]
```diff
@@ -2823,6 +2869,518 @@ def reseed(self,seed):
...
+      what = options['what'].split(",")
+      if what[0].strip().lower() == 'all':
+        what = None
+    # Target
+    target = options.get('Target',self.target[-1])
+    toAdd = ['expTerms','coeffRegressor','initialScaling']
+    if self.polyExpParams['coeffRegressor'].strip() == 'poly':
+      toAdd.append('polyOrder')
+    for add in toAdd:
+      if (add in what) if what is not None else (True):
```
Comment (mandd) [inline framework/SupervisedLearning.py]: this line totally makes no sense to me

File: framework/SupervisedLearning.py [hunk 5]
```diff
@@ -2823,6 +2869,518 @@ def reseed(self,seed):
...
+      @ In, names, unused
+      @ In, feat, feature to normalize
+      @ Out, None
+    """
+    self.muAndSigmaFeatures[feat] = (0.0,1.0)
+
+  #######
+  def __getTimeScale(self,dmd=True):
+    """
+      Get the ts of the dmd (if dmd = True) or training (if dmd = False) reconstructed time scale.
```
Comment (mandd) [inline framework/SupervisedLearning.py]: ts? timescale?

File: framework/SupervisedLearning.py [hunk 6]
```diff
@@ -2823,6 +2869,518 @@ def reseed(self,seed):
...
+    self._eigs, self._modes = mathUtils.computeEigenvaluesAndVectorsFromLowRankOperator(self.__Atilde, Y, U, s, V, self.dmdParams['exactModes'])
+    if imposedHODMD:
+      self._modes = self._modes[:targetVals[:,:,targetParamIndex].shape[0], :]
+    self._amplitudes = mathUtils.computeAmplitudeCoefficients(self._modes, targetVals[:,:,targetParamIndex]/self.dmdParams['initialScaling'],self._eigs, self.dmdParams['optimized'])
+    # Default timesteps (even if the time history is not equally spaced in time, we "trick" the dmd to think it).
+    self.trainingTimeScale = {'t0': 0, 'intervals': nTimeSteps - 1, 'dt': 1}
+    self.dmdTimeScale      = {'t0': 0, 'intervals': nTimeSteps - 1, 'dt': 1}
+
+  def __evaluateLocal__(self,featureVals):
+    """
```
Comment (mandd) [inline framework/SupervisedLearning.py]: missing description

File: framework/SupervisedLearning.py [hunk 7]
```diff
@@ -2823,6 +2869,518 @@ def reseed(self,seed):
...
+    if self.dmdParams['energyRankSVD'] is not None:
+      toAdd.append('energyRankSVD')
+    else:
+      toAdd.append('rankSVD')
+      self.dmdParams['rankSVD'] = self.dmdParams['rankSVD'] if self.dmdParams['rankSVD'] is not None else -1
+
+    for add in toAdd:
+      if (add in what) if what is not None else (True):
+        outFile.addScalar(target,add,self.dmdParams[add])
+    if ("features" in what) if what is not None else (True):
```
Comment (mandd) [inline framework/SupervisedLearning.py]: again with this convoluted lines......

File: framework/SupervisedLearning.py [hunk 8]
```diff
@@ -2864,3 +3424,35 @@ def returnClass(ROMclass,caller):
     return __interfaceDict[ROMclass]
   except KeyError:
     caller.raiseAnError(NameError,'not known '+__base+' type '+ROMclass)
+
+
+def constructAB():
```
Comment (mandd) [inline framework/SupervisedLearning.py]: docstring missing

File: framework/SupervisedLearning.py [hunk 9]
```diff
@@ -2864,3 +3424,35 @@ def returnClass(ROMclass,caller):
...
+  default = False
+  if default:
+    aij   = np.zeros( (48, 10))
+    bij   = np.zeros((48, 10))
+    aij[:,0]=[1.3095000E-04,1.1592000E-04,8.0293459E-05,5.9641824E-05,1.8282000E-04,3.0850000E-04,8.4467738E-05,2.1216000E-04,9.0819652E-05,8.6864550E-05,1.4170000E-04,5.8777881E-05,9.5372655E-05,1.0809000E-04,8.5151959E-05,1.4011000E-04,7.6851612E-05,6.8748036E-05,9.1416271E-05,2.0597000E-04,3.2060000E-04,3.0995000E-04,1.8275000E-04,1.3413000E-04,9.1381893E-05,1.5891000E-04,2.7617000E-04,2.3555000E-04,3.4894000E-04,7.6148659E-05,6.2454447E-05,1.5608000E-04,1.3246000E-04,1.1963000E-04,1.8625000E-04,1.5455000E-04,3.3168000E-04,4.5678000E-04,6.6706768E-05,4.8624431E-05,2.6367000E-04,1.0008000E-04,9.7437945E-05,7.0109587E-05,1.1058000E-04,2.2267000E-04,9.0585234E-05,9.0198248E-05]
```
Comment (mandd) [inline framework/SupervisedLearning.py]: why we need to store all these values?

File: framework/utils/utils.py [hunk 10]
```diff
@@ -394,7 +395,7 @@ def isString(s):
     @ In, s, object, variable to be tested.
```
Comment (mandd) [inline framework/utils/utils.py]: docstring seems not matching the code

File: framework/SupervisedLearning.py (hunk 6 above)
Comment (alfoa) [reply framework/SupervisedLearning.py]: added

File: framework/SupervisedLearning.py (hunk 8 above)
Comment (alfoa) [reply framework/SupervisedLearning.py]: removed method

File: framework/SupervisedLearning.py (hunk 9 above)
Comment (alfoa) [reply framework/SupervisedLearning.py]: removed

File: framework/utils/utils.py (hunk 10 above)
Comment (alfoa) [reply framework/utils/utils.py]: modified

**Comment by alfoa** [comment]
@mandd I addressed your comments.
In addition, I added the tests for DMD and PolyExponential and added the possibility of the multitarget option

**Comment by mandd** [comment]
Preparing to merge

**Pull request merged**